lib/tasks/demo.rake

```diff
@@ -165,21 +197,42 @@ namespace :demo do
     # Remove analytics that are created in admin:import_or_update_jurisdictions task
     Analytic.delete_all
 
+    limit = (ENV['LIMIT'] || 1_500_000).to_i
     days = (ENV['DAYS'] || 14).to_i
     num_patients_today = (ENV['COUNT'] || 25).to_i
     cache_analytics = (ENV['SKIP_ANALYTICS'] != 'true')
 
     jurisdictions = Jurisdiction.all
-    assigned_users = Hash[jurisdictions.pluck(:id).map { |id| [id, 10.times.map { |n| Faker::Number.number(digits: 6) }] }]
+    assigned_users_range = (1..9_999).to_a.freeze
```

Comment:
Is this a question or a comment?

Reply:
Are you saying this needs to be updated to support the 6-digit assigned user?

Reply:
I think this PR might have been created back when it was 4 digits and was updated to 6 digits since then, I also changed the implementation to use `Faker` to generate the `assigned_users` number instead of instantiating an array of 100,000 integers. I guess an alternative could also be to write it as:

```
assigned_users = Hash[jurisdictions.pluck(:id).map { |id| [id, 10.times.map { rand(1..100_000) }] }]
```

Reply:
https://stackoverflow.com/questions/4395095/how-to-generate-a-random-number-between-a-and-b-in-ruby

Reply:
I benchmarked a few different approaches and got these results:

```ruby
# frozen_string_literal: true
require 'benchmark/ips'

module Benchmarking
  @jurisdictions = Jurisdiction.all

  def rand_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { rand(1..90_000) }] }]
  end

  def hash_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { |n| Faker::Number.number(digits: 6) }] }]
  end

  def freeze_assigned_user
    assigned_users_range = (1..9_999).to_a.freeze
    assigned_users = Hash[@jurisdictions.pluck(:id).map {|id| [id, assigned_users_range]}]
  end

  def benchmark
    Benchmark.ips do |x|
      x.report('rand_assigned_user') { rand_assigned_user }
      x.report('hash_assigned_user') { hash_assigned_user }
      x.report('freeze_assigned_user') { freeze_assigned_user }
      x.compare!
    end
  end
end
```

`include Benchmarking; Benchmarking.benchmark`

```
Comparison:
freeze_assigned_user:     2024.5 i/s
  rand_assigned_user:     1932.4 i/s - same-ish: difference falls within error
  hash_assigned_user:        9.3 i/s - 216.92x  (± 0.00) slower
```

Reply:
@jkufro can you run these with the range on all of them being the same? 

```suggestion
    assigned_users_range = (1..999_999).to_a.freeze
```

Even if the `freeze_assigned_user` method is slightly faster, I don't really like the idea of allocating an array of a million numbers in memory especially when it's not that necessary. The `hash_assigned_user` method though does seem to be significantly slower and I think the `rand_assigned_user` is probably the way to go.

Reply:
```ruby
# frozen_string_literal: true
require 'benchmark/ips'

module Benchmarking
  @jurisdictions = Jurisdiction.all

  def rand_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { rand(1..999_000) }] }]
  end

  def hash_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { |n| Faker::Number.number(digits: 6) }] }]
  end

  def freeze_assigned_user
    assigned_users_range = (1..999_999).to_a.freeze
    assigned_users = Hash[@jurisdictions.pluck(:id).map {|id| [id, assigned_users_range]}]
  end

  def benchmark
    Benchmark.ips do |x|
      x.report('rand_assigned_user') { rand_assigned_user }
      x.report('hash_assigned_user') { hash_assigned_user }
      x.report('freeze_assigned_user') { freeze_assigned_user }
      x.compare!
    end
  end
end
```

```
Comparison:
freeze_assigned_user:       30.8 i/s
  rand_assigned_user:        8.2 i/s - 3.75x  (± 0.00) slower
  hash_assigned_user:        0.0 i/s - 1823.01x  (± 0.00) slower
```

Reply:
I tried another method `rand_assigned_user_dup_list`. I'm good with changing to that - I don't think we really need the full freeze array and having 10,000 possibilities is good enough IMO

```ruby
# frozen_string_literal: true
require 'benchmark/ips'

module Benchmarking
  @jurisdictions = Jurisdiction.all

  def rand_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { rand(1..999_000) }] }]
  end

  def rand_assigned_user_dup_list
    assigned_users_range = 10_000.times.map { rand(1..999_000) }
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, assigned_users_range]}]
  end

  def hash_assigned_user
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, 200.times.map { |n| Faker::Number.number(digits: 6) }] }]
  end

  def freeze_assigned_user
    assigned_users_range = (1..999_999).to_a.freeze
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, assigned_users_range]}]
  end

  def benchmark
    Benchmark.ips do |x|
      # x.report('rand_assigned_user') { rand_assigned_user }
      x.report('rand_assigned_user_dup_list') { rand_assigned_user_dup_list }
      # x.report('hash_assigned_user') { hash_assigned_user }
      x.report('freeze_assigned_user') { freeze_assigned_user }
      x.compare!
    end
  end
end
```

```
Comparison:
rand_assigned_user_dup_list:      136.9 i/s
freeze_assigned_user:       30.0 i/s - 4.56x  (± 0.00) slower
```

Reply:
@timwongj @Bialogs are we ok to resolve this thread?

Reply:
I like the idea, we can probably even try

```
def rand_assigned_user_dup_list
    assigned_users_range = 1_000.times.map { rand(1..999_000) }
    assigned_users = Hash[@jurisdictions.pluck(:id).map { |id| [id, assigned_users_range]}]
  end
```

to save memory

Reply:
I reduced the size to 1000

Reply:
Sorry, I actually just remembered why I wrote it the way I did months ago like this:
```
assigned_users = Hash[jurisdictions.pluck(:id).map { |id| [id, 10.times.map { |n| Faker::Number.number(digits: 6) }] }]
```

The reason for not freezing `assigned_users_range` and creating a different assigned users range (with only 10 or so different possibilities per jurisdiction) is to more closely mimic the way the `assigned_users` field is used in the context of this app. Jurisdictions typically use this field to assign/group certain patients to different users within their jurisdiction (without explicitly strictly mapping patients to a user id).

For that reason, we would like the variety of this sample data to reflect that by having a more limited set of numbers for assigned users (still 1-999,999 though) to choose from, with those sets of numbers being different for each jurisdiction.

This is what the numbers look like with `rand_assigned_user_dup_list`
```
2.7.3 :032 > Jurisdiction.all.map { |jur| [Patient.where(jurisdiction_id: jur.id).size, Patient.where(jurisdiction_id: jur.id).distinct.pluck(:assigned_user).size] }
 => [[114, 84], [131, 98], [123, 90], [129, 100], [142, 101], [143, 114], [159, 129]] 
```

and this is what it looked like before:
```
2.7.3 :032 > Jurisdiction.all.map { |jur| [Patient.where(jurisdiction_id: jur.id).size, Patient.where(jurisdiction_id: jur.id).distinct.pluck(:assigned_user).size] }
 => [[151, 20], [123, 17], [145, 22], [139, 20], [147, 18], [130, 16], [144, 20]] 
```

The difference between the two is that there's significantly more variety in the number of distinct assigned users per jurisdiction. The old implementation can certainly be improved though, and I'm thinking it can be something like this to avoid using `Faker`:
```
assigned_users = Hash[jurisdictions.pluck(:id).map { |id| [id, 10.times.map { rand(1..100_000) }] }]
```
```
2.7.3 :032 > Jurisdiction.all.map { |jur| [Patient.where(jurisdiction_id: jur.id).size, Patient.where(jurisdiction_id: jur.id).distinct.pluck(:assigned_user).size] }
 => [[141, 21], [157, 16], [126, 15], [136, 17], [132, 20], [158, 14], [191, 28]] 
```

Reply:
@timwongj just pushed that  suggested update

Reply:
I just added a commit to include a comment explaining what we just talked about